File: 2025/day04-1.hs
type Paper = (Int, Int)

solution :: String -> Int
solution content = length . filter (< 4) . map (length . neighbors papers) $ papers
  where
    papers = parse (0, 0) content

neighbors :: [Paper] -> Paper -> [Paper]
neighbors ps (x, y) = take 4 . filter (`elem` [(x - 1, y), (x + 1, y), (x, y - 1), (x, y + 1), (x - 1, y - 1), (x - 1, y + 1), (x + 1, y - 1), (x + 1, y + 1)]) $ ps

parse :: Paper -> String -> [Paper]
parse _ [] = []
parse (x, y) ('@' : xs) = (x, y) : parse (x + 1, y) xs
parse (x, y) ('\n' : xs) = parse (0, y + 1) xs
parse (x, y) (_ : xs) = parse (x + 1, y) xs

main :: IO ()
main = do readFile "day04.dat" >>= print . solution

File: 2025/day04_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type Coords = tuple[int, int]


def solution(content: str) -> Any:
    rolls = parse(content)
    return len([roll for roll in rolls if is_accessible(rolls, roll)])


def is_accessible(rolls: frozenset[Coords], roll: Coords) -> bool:
    row, col = roll
    return (
        len(
            [
                r
                for r in (
                    (row - 1, col - 1),
                    (row - 1, col),
                    (row - 1, col + 1),
                    (row, col - 1),
                    (row, col + 1),
                    (row + 1, col - 1),
                    (row + 1, col),
                    (row + 1, col + 1),
                )
                if r in rolls
            ]
        )
        < 4
    )


def parse(content: str) -> frozenset[Coords]:
    return frozenset(
        (row_index, col_index)
        for row_index, row in enumerate(content.split("\n"))
        for col_index, col in enumerate(row)
        if col == "@"
    )


def main() -> None:
    with open("day04.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()
